lib/Sema/CSSolver.cpp

```diff
@@ -1358,6 +1358,224 @@ static Constraint *selectBestBindingDisjunction(
   return firstBindDisjunction;
 }
 
+// For a given type, determine if it's either a concrete type, the
+// result of converting from a concrete type, or a type variable known
+// to conform to other types (or the result of converting from such a
+// type variable).
+static bool havePotentialTypesOrLiteralConformances(Type ty,
+                                                    ConstraintSystem &cs) {
+  llvm::SmallSet<TypeVariableType *, 4> visited;
+  llvm::SmallVector<Type, 4> worklist;
+  worklist.push_back(ty);
+
+  while (!worklist.empty()) {
+    auto itemTy = worklist.pop_back_val()->getWithoutSpecifierType();
+
+    if (!itemTy->is<TypeVariableType>())
+      return true;
+
+    auto tyvar = itemTy->castTo<TypeVariableType>();
+    if (cs.getFixedType(tyvar))
+      return true;
+
+    auto *rep = cs.getRepresentative(tyvar);
+
+    // FIXME: This can happen when we have two type variables that are
+    // subtypes of each other. We would ideally merge those type
+    // variables somewhere.
+    if (visited.count(rep))
+      continue;
+
+    visited.insert(rep);
+
+    // Gather all the constraints involving this type variable, and
+    // then attempt to trace back through each constraint to see if we
+    // can reach a concrete type or a literal.
+
+    llvm::SetVector<Constraint *> constraints;
+    cs.getConstraintGraph().gatherConstraints(
+        rep, constraints, ConstraintGraph::GatheringKind::EquivalenceClass,
+        [](Constraint *constraint) {
```

Comment:
Nit: default callback already does `return true;` so you don't have to explicitly provide it...